### ISSUE TITLE
unittest.subTest and way to selectively mark subTests as expected failures

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#82669]
### 🚀 The feature, motivation and pitch

When testing if something like vmap works on an operator, we test the following things:
1. if it errors out (it shouldn't)
2. if the output of vmap matches the output of a for loop (it should)
3. if there is a batching rule implemented for the operation. We do this by running the vmap and checking if it raises any "batching rule not implemented" warnings.

We have two separate tests, test_vmap_exhaustive, and test_op_has_batch_rule. The former does (1) and (2), and the latter does (1), (2), (3) (because (1) and (2) are almost required for (3)). We could cut down the test time if we used something like unittest.subTest, which allows people to define smaller tests inside a single test.

Furthermore, if a vmap test fails for e.g. torch.searchsorted, we have an expected failure for it. It would be nice to be able to distinguish between if it failed because of (1) or if it failed because of (2); (2) is a silent correctness issue and is much more hi-pri to fix.

### Alternatives

n/a

### Additional context

cc @jbschlosser who has worked on many testing improvements in the past

**Comment by jbschlosser** [comment]
One possible design for this:

* Define a `subTest()` context manager that stores identification data for the subtest (e.g. `with subTest(part='batching_rule_check'):`)
* Define a `SubTestException` that consists of an exception and identification data for the subtest
    * `subTest()` should catch any exception that occurs inside and re-throw in the form of a `SubTestException`, additionally including the subtest identification data
* Define an `expectedSubTestFailure(**subtest_data)` decorator that specially handles `SubTestException`, only swallowing the exception if `subtest_data` matches the subtest_data in the caught exception
    * If the subtest_data doesn't match or no exception is thrown, the decorator fails the test

Example all-green success case:
```python
class TestFoo(TestCase):
    @expectedSubTestFailure(part='for_loop_check')
    def test_something(self):
        with subTest(part='error_check'):
            pass
        with subTest(part='for_loop_check'):
            self.fail('boom')
        with subTest(part='batching_rule_check'):
            pass
```

Fails test with unexpected success:
```python
class TestFoo(TestCase):
    @expectedSubTestFailure(part='for_loop_check')
    def test_something(self):
        with subTest(part='error_check'):
            pass
        with subTest(part='for_loop_check'):
            pass
        with subTest(part='batching_rule_check'):
            pass
```

Fails test with unexpected success AND exception stacktrace from the `batching_rule_check` subtest:
```python
class TestFoo(TestCase):
    @expectedSubTestFailure(part='for_loop_check')
    def test_something(self):
        with subTest(part='error_check'):
            pass
        with subTest(part='for_loop_check'):
            pass
        with subTest(part='batching_rule_check'):
            raise RuntimeError('boom')
```

To support `@expectedSubTestFailure` for only certain cases of parametrized tests, I think we'd need to finish up #79979 and add a decorator that conditionally applies a given decorator for certain param values.

@zou3519 wdyt about this? I have a PoC working locally that seems correct.

**Comment by zou3519** [comment]
Thanks for the quick response, Joel, I thought this was going to be more difficult. API seems reasonable to me. Does it compose with OpInfo testing?

A question about semantics. If we have the following:
```py
class TestFoo(TestCase):
    @expectedSubTestFailure(part='for_loop_check')
    def test_something(self):
        with subTest(part='error_check'):
            # A
            pass
        with subTest(part='for_loop_check'):
            # B
            pass
        with subTest(part='batching_rule_check'):
            # C
            pass
```
Then does C end up executing? (My guess is no, because C may depend on variables created in B)

**Comment by zou3519** [comment]
Another question is, does this play nice with unittest test reporting? Unittest has some nice error reporting for subtests (https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests), though they don't have a concept of expected failure on specific subtests

**Comment by jbschlosser** [comment]
> Does it compose with OpInfo testing?

Yes, I expect it to compose well with OpInfo testing. You could use the decorator within a `DecorateInfo` context to do per-op flagging of certain subtests as expected failures.

> A question about semantics. If we have the following:
```python
class TestFoo(TestCase):
    @expectedSubTestFailure(part='for_loop_check')
    def test_something(self):
        with subTest(part='error_check'):
            # A
            pass
        with subTest(part='for_loop_check'):
            # B
            pass
        with subTest(part='batching_rule_check'):
            # C
            pass
```
> Then does C end up executing? (My guess is no, because C may depend on variables created in B)

For the impl outlined above, usage of `@expectedSubTestFailure` wouldn't affect the test logic's flow control. If B runs without error, C will run. If B errors out / fails, C will not run, just like in normal Python. After the test logic is run to either completion or error as usual, the decorator will kick in and fail the test if it didn't see the subtest failure it expected.

> Another question is, does this play nice with unittest test reporting? Unittest has some nice error reporting for subtests (https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests), though they don't have a concept of expected failure on specific subtests

The impl I described somewhat reimplements the `unittest.subTest` mechanism so that we can include the concept of expected failure on specific subtests. Because of this, we unfortunately wouldn't get the nice subtest-based reporting from unittest. The overall test would still be treated as a single entity for reporting purposes. That said, we should be able to include info on which subTest was being executed as part of the test failure message.

Any concerns with the above? @zou3519

**Comment by zou3519** [comment]
> Any concerns with the above? @zou3519

The above all SGTM.

I'm a bit sad that we don't get the nice subtest-based reporting from unittest (or the nice expected failure reporting from unittest/pytest), but I don't think we can do anything about that without writing our own test runner

**Comment by zou3519** [comment]
@jbschlosser I'm attempting to refactor the vmap tests and I noticed something interesting.

In the ideal case we would want to be able to xfail each individual OpInfo using a expectedSubTestFailure using error_check, for_loop_check, or batching_rule_check.

The natural way is to write the tests like the following, 
```py
@ops(op_db)
def test_vmap(self, device, dtype, op):
  for sample_input in sample_inputs:
    with subTest(part='error_check'):
      with capture_warnings as w:
        output = vmap(...)
    with subTest(part='for_loop_check'):
      expected_output = for_loop_check(...)
      assert output == expected_output
    with subTest(part='batching_rule_check'):
      assert w is None
```
But this creates `len(sample_inputs) * 3` subtests, which is not what we want here. So we would end up writing the tests like the following to accomplish our goal of being able to label each OpInfo as either failing the error_check, for_loop_check, or batching_rule_check:

```py
@ops(op_db)
def test_vmap(self, device, dtype, op):
  outputs = []
  expected_outputs = []
  has_batching_rule = []
  with subTest(part='error_check'):
    for i, sample_input in enumerate(sample_inputs):
      with capture_warnings() as w:
        outputs[i] = vmap(...)
      has_batching_rule[i] = (w is None)
  with subTest(part='for_loop_check'):
    for i, sample_input in enumerate(sample_inputs):
      expected_outputs[i] = ...
      assert outputs[i] == expected_outputs
  with subTest(part='batching_rule_check'):
    assert all(has_batching_rule)
```
The downside of this approach is that we need to thread state through the subtests. This seems not ideal, because it makes it more difficult to debug: `outputs[i]` was computed in a different for-loop from `expected_outputs[i]`, so when using pdb you'd need to set two breakpoints instead of just one

I don't think anything can be really done about this, this comment is more of a complaint

**Comment by jbschlosser** [comment]
@zou3519 Interesting - assuming the implementation I described above, I think your first example would actually give you the behavior you want.

```python
@ops(op_db)
def test_vmap(self, device, dtype, op):
  for sample_input in sample_inputs:
    with subTest(part='error_check'):
      with capture_warnings as w:
        output = vmap(...)
    with subTest(part='for_loop_check'):
      expected_output = for_loop_check(...)
      assert output == expected_output
    with subTest(part='batching_rule_check'):
      assert w is None
```

Rather than thinking of this as creating `len(sample_inputs) * 3` subtests, you can think of this as entering and leaving the same 3 "subtest contexts" `len(sample_inputs)` times each.

If you in fact wanted `len(sample_inputs) * 3` distinct subtests, that could look like this (note the use of `sample_index` to create unique subtest contexts). This would allow you to exFail a specific sample input for a specific check (possibly not a good idea since this assumes the sample order, but anyway):

```python
@ops(op_db)
def test_vmap(self, device, dtype, op):
  for i, sample_input in enumerate(sample_inputs):
    with subTest(part='error_check', sample_index=i):
      with capture_warnings as w:
        output = vmap(...)
    with subTest(part='for_loop_check', sample_index=i):
      expected_output = for_loop_check(...)
      assert output == expected_output
    with subTest(part='batching_rule_check', sample_index=i):
      assert w is None
```

A subtest context is defined by the args passed to `subTest`, and you have full control over how granular you want them to be.

**Comment by zou3519** [comment]
> Rather than thinking of this as creating len(sample_inputs) * 3 subtests, you can think of this as entering and leaving the same 3 "subtest contexts" len(sample_inputs) times each.

Thanks for the explanation. I think I've grokked the implementation. To check my understanding, the following would pass. Removal of either of the `expectedSubTestFailure(part='batching_rule_check')` would cause a failure (because =the error wouldn't be swallowed anymore)


```py
class TestFoo(TestCase):
    @expectedSubTestFailure(part='for_loop_check')
    @expectedSubTestFailure(part='batching_rule_check')
    def test_something(self):
        with subTest(part='error_check'):
            pass
        with subTest(part='for_loop_check'):
            pass
        with subTest(part='batching_rule_check'):
            raise RuntimeError('boom')

        with subTest(part='error_check'):
            pass
        with subTest(part='for_loop_check'):
            raise RuntimeError('boom')
        with subTest(part='batching_rule_check'):
            pass
 ```
